pkg/controllers/vdb/stopsubcluster_reconciler.go

```diff
@@ -0,0 +1,156 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package vdb
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	vapi "github.com/vertica/vertica-kubernetes/api/v1"
+	"github.com/vertica/vertica-kubernetes/pkg/cmds"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin/opts/stopsc"
+	config "github.com/vertica/vertica-kubernetes/pkg/vdbconfig"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type StopSubclusterReconciler struct {
+	Rec            config.ReconcilerInterface
+	Vdb            *vapi.VerticaDB // Vdb is the CRD we are acting on.
+	PRunner        cmds.PodRunner
+	Log            logr.Logger
+	PFacts         *PodFacts
+	InitiatorPodIP string
+	Dispatcher     vadmin.Dispatcher
+}
+
+func MakeStopSubclusterReconciler(rec config.ReconcilerInterface, log logr.Logger,
+	vdb *vapi.VerticaDB, pfacts *PodFacts, dispatcher vadmin.Dispatcher) controllers.ReconcileActor {
+	return &StopSubclusterReconciler{
+		Rec:        rec,
+		Log:        log.WithName("StopSubclusterReconciler"),
+		Vdb:        vdb,
+		PFacts:     pfacts,
+		Dispatcher: dispatcher,
+	}
+}
+
+func (s *StopSubclusterReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	err := s.PFacts.Collect(ctx, s.Vdb)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	scsToStop := s.findSubclustersWithShutdownNeeded()
+	if len(scsToStop) == 0 {
+		s.Log.Info("Aborting stop subcluster. No subclusters needing shutdown were found.")
+		return ctrl.Result{}, nil
+	}
+	oldSize := len(scsToStop)
+	scsToStop = s.filterSubclustersWithRunningPods(scsToStop)
+	needRequeue := false
+	if len(scsToStop) != oldSize {
+		needRequeue = true
+	}
+
+	if err := s.stopSubclusters(ctx, scsToStop); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if needRequeue {
+		s.Log.Info("Some subclusters that need shutdown have pods not running. Requeue reconciliation.")
+	}
+	return ctrl.Result{Requeue: needRequeue}, nil
+}
+
+// getFirstUpSCPodIP finds and return the first up node in the given
+// subcluster
+func (s *StopSubclusterReconciler) getFirstUpSCPodIP(scName string) (string, bool) {
+	return s.PFacts.FindFirstUpPodIP(true, scName)
+}
+
+// stopSubclusters call the stop subcluster api call on the given subclusters
+func (s *StopSubclusterReconciler) stopSubclusters(ctx context.Context, scs []string) error {
+	for _, sc := range scs {
+		// A subcluster needs to be shutdown only if at least
+		// one node is up. So we check for an up node and if we find
+		// one, we use it as initiator for stop subcluster, if we don't
+		// then we skip the current subcluster and move to next one
+		ip, ok := s.getFirstUpSCPodIP(sc)
+		if !ok {
+			continue
```

Comment:
You verified the running pods in filterSubclustersWithRunningPods(). If we come here and the sc has any pod stopped, we'd better requeue the reconciler since 'needRequeue' is not accurate this time. You calculated 'needRequeue' based on the result of filterSubclustersWithRunningPods().